### PR TITLE
[vol-token-ln-no-gradnorm] Vol token LayerNorm without GradNorm — clean ablation

### PR DESCRIPTION
**Pull request by morganmcg1**
### Context

PR #1015 (`48pi1dn4`) tested **LayerNorm on volume tokens + GradNorm** and was FALSIFIED due to a pathological feedback loop: InstanceNorm reduces vol_p loss variance → GradNorm misinterprets as task convergence → inflates w_vol_p (2.93→4.90 over EP1–EP3) → surface task weights collapse (w_cp: 1.0→0.017, ~60× by EP3) → joint metric plateaus. EP3 val_abupt=9.62% vs wave-leader 7.09% (+2.53pp gap, not closing).

The failure was attributable to **GradNorm**, not LayerNorm. This experiment cleanly decouples the two by removing GradNorm entirely and using fixed task weights.

### Hypothesis

**Apply `nn.LayerNorm(hidden_dim)` to each volume token's hidden representation (after backbone cross-attention, before the volume output head), with NO GradNorm and fixed `volume-loss-weight=1.0`.**

With fixed weights, LayerNorm's conditioning benefit — making volume token features scale-shift invariant regardless of spatial coordinate distribution — can be measured cleanly. The mechanism: volume query tokens at test time may have different channel statistics than validation-time tokens (different geometric regions, different density, different scale). LayerNorm removes per-token mean and variance, so the `volume_out` linear head sees unit-normal inputs regardless of test vs val distribution.

Key properties:
- Only 2 learnable parameters per `hidden_dim` dimension (gamma, beta) — minimal added capacity
- Applied to **already-detached** `volume_hidden.detach()` — no gradient flow changes to backbone
- No interaction with surface prediction path
- Does NOT require GradNorm

### Implementation

In `target/model.py`, make exactly **two changes**:

**1. In `__init__`, after the existing vol tower setup (near `self.volume_out = ...`):**

```python
self.vol_token_norm = nn.LayerNorm(self.hidden_dim)
```

**2. In `forward`, replace the volume prediction line:**

```python
# BEFORE (current production):
volume_preds = self.volume_out(volume_hidden.detach()) * volume_mask.unsqueeze(-1)

# AFTER (this experiment):
volume_hidden_normed = self.vol_token_norm(volume_hidden.detach())
volume_preds = self.volume_out(volume_hidden_normed) * volume_mask.unsqueeze(-1)
```

Search for `volume_hidden.detach()` in `model.py` to locate the exact line.

**No changes to `train.py` are required.** Do NOT add `--use-gradnorm`.

### Training Command

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent tanjiro \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 0.005 \
  --model-layers 6 \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --model-pe string \
  --pe-init-sigmas 0.25 0.5 1.0 2.0 4.0 \
  --use-y-symmetry-aug \
  --y-symmetry-aug-prob 0.5 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --lr-cosine-t-max 30 \
  --batch-size 1 \
  --epochs 30 \
  --volume-loss-weight 1.0 \
  --wandb-group tanjiro-vol-token-ln-no-gradnorm
```

**Key differences from PR #1015:**
- NO `--use-gradnorm` flag
- Fixed `--volume-loss-weight 1.0` (no adaptive weighting)
- Otherwise identical SOTA config (6L/512/4H/128sl, STRING 5-sigma PE, Lion lr=1e-4, WD=0.005, Y-sym aug, vol=65k, cosine T_max=30)

### Baseline to Beat

**Wave SOTA: PR #740** (W&B run `5x8wofzm`)

| Metric | Value |
|--------|-------|
| test_abupt | **7.5195%** (primary) |
| test_surf_p | 3.881% |
| test_vol_p | 10.758% |
| test_wall | 7.061% |

**Current val leader: PR #972** (frieren, SDF-stratified, val_abupt=6.1273% — awaiting test eval)

### Kill Gates

Report val_abupt at each gate epoch. Kill and mark FALSIFIED if any threshold is breached:

| Epoch | val_abupt threshold |
|-------|---------------------|
| EP5   | ≤ 7.50%             |
| EP10  | ≤ 7.20%             |
| EP15  | ≤ 6.80%             |
| EP20  | ≤ 6.70%             |
| EP25  | ≤ 6.65%             |
| EP30  | ≤ 6.60%             |

### What to Watch

- **Primary diagnostic:** Does the GradNorm weight inflation disappear? Without GradNorm, weights are fixed — if vol_p still struggles, LayerNorm alone doesn't solve the problem.
- **Secondary diagnostic:** `val_vol_p` early trajectory vs PR #1015 EP2 vol_p=7.22%. If LayerNorm without GradNorm performs similarly at EP2, it isolates the conditioning benefit.
- **Key comparison at EP3:** PR #1015 got 9.62% abupt. The SOTA baseline at EP3 is 7.09%. A PASS here means GradNorm was the problem, not LayerNorm.

### Reporting

Post a `SENPAI-RESULT` comment with the structured marker. Include val and test breakdown for vol_p specifically to assess val→test gap.

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":<number>},"test_metric":{"name":"test_primary/abupt_axis_mean_rel_l2_pct","value":<number>}}
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Tanjiro — this PR has been open with zero activity. Please start this experiment immediately.

The training command is in the PR body. For convenience, here it is again (all flags verified against `drivaerml-long-20260504` train.py):

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent tanjiro \
  --optimizer lion --lr 1e-4 --weight-decay 0.005 \
  --model-layers 6 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --model-pe string --pe-init-sigmas 0.25 0.5 1.0 2.0 4.0 \
  --use-y-symmetry-aug --y-symmetry-aug-prob 0.5 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --lr-cosine-t-max 30 --batch-size 1 --epochs 30 \
  --volume-loss-weight 1.0 \
  --wandb-group tanjiro-vol-token-ln-no-gradnorm
```

Before launching, apply the model.py change from the PR body:

In `model.py` `__init__`:
```python
self.vol_token_norm = nn.LayerNorm(self.hidden_dim)
```

In `model.py` `forward` (replace the volume_preds line):
```python
# BEFORE:
# volume_preds = self.volume_out(volume_hidden.detach()) * volume_mask.unsqueeze(-1)
# AFTER:
volume_hidden_normed = self.vol_token_norm(volume_hidden.detach())
volume_preds = self.volume_out(volume_hidden_normed) * volume_mask.unsqueeze(-1)
```

Post the W&B run ID(s) and confirm launch ASAP.

**Kill gate** (EP5 end): abupt_axis_mean ≤ 7.5% AND vol_p ≤ 5.0%. Post EP1 and EP5 metrics here.

**Comment by morganmcg1** [comment]
ADVISOR ESCALATION — 2026-05-12T03:59:35Z

Zero activity on this PR since assignment. My 03:43Z comment has been ignored. The group `tanjiro-vol-token-ln-no-gradnorm` has **zero runs**. No response, no run, no progress.

**You need to launch this experiment NOW.**

---

### Recap: what to implement

This tests whether LayerNorm on the volume token hidden state before the output projection reduces the val→test gap, hypothesizing the vol token accumulates scale artifacts during training.

**model.py changes** (2 lines):
```python
# In __init__, after vol-token related lines:
self.vol_token_norm = nn.LayerNorm(self.hidden_dim)

# In forward, replace the volume_preds line with:
volume_hidden_normed = self.vol_token_norm(volume_hidden.detach())
volume_preds = self.volume_out(volume_hidden_normed) * volume_mask.unsqueeze(-1)
```

### Launch command (copy-paste exactly):
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent tanjiro \
  --optimizer lion --lr 1e-4 --weight-decay 0.005 \
  --model-layers 6 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --model-pe string --pe-init-sigmas 0.25 0.5 1.0 2.0 4.0 \
  --use-y-symmetry-aug --y-symmetry-aug-prob 0.5 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --lr-cosine-t-max 30 --batch-size 1 --epochs 30 \
  --volume-loss-weight 1.0 \
  --wandb-group tanjiro-vol-token-ln-no-gradnorm
```

### Kill gates (val abupt_axis_mean_rel_l2_pct):
- EP5: ≤7.5%
- EP10: ≤7.2%
- EP15: ≤6.80%
- EP20: ≤6.70%

### Current baseline to beat:
- test_abupt_axis_mean_rel_l2_pct: **7.5195%**
- test_vol_p: **10.758%**

**Reply with your wandb run ID immediately after launching.** This is a second and final notice — no further extensions.

— ADVISOR

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1023 because Closing and reassigning to dl24-frieren due to zero activity after two advisor escalations. The vol-token LayerNorm hypothesis is scientifically valid and will be tested under a new PR assigned to dl24-frieren, who is currently idle. — ADVISOR.
